bitcoin/src/psbt/mod.rs

```diff
@@ -72,87 +283,480 @@ impl Psbt {
-                (Some(witness_utxo), _) => Ok(witness_utxo),
-                (None, Some(non_witness_utxo)) => {
-                    let vout = tx_input.previous_output.vout as usize;
-                    non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
-                }
-                (None, None) => Err(Error::MissingUtxo),
+    pub fn iter_funding_utxos(&self) -> Box<dyn Iterator<Item = Result<&TxOut, Error>> + '_> {
+        let inner = &self.inner;
+        match inner.version {
+            Version::PsbtV0 => {
+                let unsigned_tx = inner.unsigned_tx.as_ref().unwrap();
+                assert_eq!(inner.inputs.len(), unsigned_tx.input.len());
+                Box::new(unsigned_tx.input.iter().zip(&inner.inputs).map(
+                    |(tx_input, psbt_input)| match (
+                        &psbt_input.witness_utxo,
+                        &psbt_input.non_witness_utxo,
+                    ) {
+                        (Some(witness_utxo), _) => Ok(witness_utxo),
+                        (None, Some(non_witness_utxo)) => {
+                            let vout = tx_input.previous_output.vout as usize;
+                            non_witness_utxo.output.get(vout).ok_or(Error::PsbtUtxoOutOfbounds)
+                        }
+                        (None, None) => Err(Error::MissingUtxo),
+                    },
+                )) as Box<dyn Iterator<Item = Result<&TxOut, Error>>>
             }
-        })
+            _ => {
+                // In PsbtV2, Input contains all the details, no need of unsigned_tx
+                Box::new(inner.inputs.iter().map(|input| {
+                    match (&input.witness_utxo, &input.non_witness_utxo) {
+                        (Some(witness_utxo), _) => Ok(witness_utxo),
+                        (None, Some(non_witness_utxo)) => {
+                            let output_index = input.output_index.unwrap() as usize;
+                            non_witness_utxo
+                                .output
+                                .get(output_index)
+                                .ok_or(Error::PsbtUtxoOutOfbounds)
+                        }
+                        (None, None) => Err(Error::MissingUtxo),
+                    }
+                })) as Box<dyn Iterator<Item = Result<&TxOut, Error>>>
+            }
+        }
     }
 
-    /// Checks that unsigned transaction does not have scriptSig's or witness data.
-    fn unsigned_tx_checks(&self) -> Result<(), Error> {
-        for txin in &self.unsigned_tx.input {
-            if !txin.script_sig.is_empty() {
-                return Err(Error::UnsignedTxHasScriptSigs);
+    /// Calculates the max lock time.
+    fn max_locktime(
+        locktime: Option<absolute::LockTime>,
+        new_lock: absolute::LockTime,
+    ) -> absolute::LockTime {
+        match locktime {
+            None => new_lock,
+            Some(lock) =>
+                if lock.is_same_unit(new_lock) && lock >= new_lock {
```

Comment:
> I'd expect calls to is_same_unit
I'd expect checking for None when using Ord (<) (or check unit before comparing)

I have added this check here as it is a local function that functions other than `add_input` and `compute_locktime` may use in the future.

But, in `add_input`, the new given `input` is validated first using `Input::validate_version()` which ensures that the `required_time_locktime` in the `input` is a `Locktime::Seconds` and `required_height_locktime` is a `Locktime::Blocks`. Similarly, the `Psbt` ensures the same for all its existing `Input`s. So there shouldn't be any situation where there is a comparison between different locktime types. Similar to the `Psbt::compute_locktime()`.

Reply:
This function returns `new_lock` if the locks are different units and there is no way for the casual reader to see why this is sane. Refactor for the win BTW, because now the anomaly jumps out at us!

Reply:
Ohh! My bad, right, sorry I missed that part.